src/hotspot/share/classfile/javaClasses.inline.hpp

```diff
@@ -293,10 +293,12 @@ inline Klass* java_lang_Class::as_Klass(oop java_class) {
 
 inline bool java_lang_Class::is_primitive(oop java_class) {
   // should assert:
-  //assert(java_lang_Class::is_instance(java_class), "must be a Class object");
+  // assert(java_lang_Class::is_instance(java_class), "must be a Class object");
   bool is_primitive = (java_class->metadata_field(_klass_offset) == nullptr);
 
 #ifdef ASSERT
+  // The heapwalker walks through Classes that have had their Klass pointers removed, so can't assert this.
+  // assert(is_primitive == java_class->bool_field(_is_primitive_offset), "must match what we told Java");
```

Comment:
I don't understand this comment about the heapwalker.  It sounds like we could have `is_primitive` set to true incorrectly.  If so, what prevents the asserts below from failing?  And why not use the value from _is_primitive_offset instead?

Reply:
> ... but not in the return since the caller likely will fetch the klass pointer next.

I notice that too.  Callers are using is_primitive() to short-circuit calls to as_Klass(), which means they seem to be aware of this implementation detail when maybe they shouldn't.

Reply:
There are 136 callers so yes, it might be something that shouldn't be known in this many places.

Reply:
Definitely out of the scope of this PR.